### PR TITLE
[flang] Add nsw flag to do-variable increment with a new option

### PR DESCRIPTION
**Pull request by yus3710-fj**
This patch adds nsw flag to the increment of do-variables when a new option is enabled.
NOTE 11.10 in the Fortran 2018 standard says they never overflow.

See also the discussion in #74709 and the following discourse post.
https://discourse.llvm.org/t/rfc-add-nsw-flags-to-arithmetic-integer-operations-using-the-option-fno-wrapv/77584/5

### CODE REVIEW

**Comment by yus3710-fj** [comment]
Windows build failed with the following message.

> fatal error C1060: compiler is out of heap space

I'll try re-running CI with an empty commit because it might have been happend by accident.

File: flang/include/flang/Lower/LoweringOptions.def [hunk 1]
```diff
@@ -34,5 +34,9 @@ ENUM_LOWERINGOPT(NoPPCNativeVecElemOrder, unsigned, 1, 0)
 /// On by default.
 ENUM_LOWERINGOPT(Underscoring, unsigned, 1, 1)
 
+/// If true, add nsw flags to arithmetic operations for integer.
+/// Off by default.
+ENUM_LOWERINGOPT(NoSignedWrap, unsigned, 1, 0)
```
Comment (vzakhari) [inline flang/include/flang/Lower/LoweringOptions.def:39]: `NoSignedWrap` sounds too generic for an option that only forces setting `NSW` for the loop variables' increments.  Please consider renaming or at least make the scope of `NSW` assignment clear in the comment above.

**Review by vzakhari — APPROVED** [review]
It looks good to me except maybe for the naming of the lowering option.  Thank you for doing this!

File: flang/include/flang/Lower/LoweringOptions.def (hunk 1 above)
Comment (yus3710-fj) [reply flang/include/flang/Lower/LoweringOptions.def]: Thank you for the review!
I changed the name of the lowering option `NoSignedWrap` to `NSWOnLoopVarInc`.

**Review by tblah — APPROVED** [review]
Looks great to me. Thanks for all of your work on this!

**Review by jeanPerier — APPROVED** [review]
Thank you!

**Pull request merged**